raw_bsd.go

```diff
@@ -192,6 +200,73 @@ func (p *packetConn) SetPromiscuous(b bool) error {
 	return syscall.SetBpfPromisc(p.fd, m)
 }
 
+// ifReq is the BSD struct passed to ioctl for multicast add/drop operations.
+type ifReq struct {
+	ifrName [unix.IFNAMSIZ]uint8
+	ifrAddr unix.SockaddrDatalink
+}
+
+const etherAdderAlen = 6
+
+// JoinMulticast joins or drops membership of data link layer multicast address on interface,
+// allowing it to recieve multicast traffic.
+func (p *packetConn) SetMulticast(b bool, addr net.HardwareAddr) error {
+	// Convert hardware address back to byte array form
+	var baddr [8]byte
+	copy(baddr[:], addr)
+
+	s, err := syscall.Socket(unix.AF_INET, unix.SOCK_DGRAM, unix.IPPROTO_UDP)
+	if err != nil {
+		return err
+	}
+	defer syscall.Close(s)
+
+	SockaddrDataLink := unix.SockaddrDatalink{
+		Len:    unix.SizeofSockaddrDatalink,
+		Family: unix.AF_LINK,
+		Index:  0,
+		Type:   0,
+		Nlen:   0,
+		Alen:   etherAdderAlen,
+		Slen:   0,
+	}
+
+	for i, v := range baddr {
+		SockaddrDataLink.Data[i] = int8(v)
+	}
+
+	ifr := ifReq{
+		ifrAddr: SockaddrDataLink,
+	}
+
+	copy(ifr.ifrName[:], p.ifi.Name)
```

Comment:
Will probably want to do a length check on the interface's name to ensure there is space for the entire name with a null terminator (I think?).